mongo-store.js

```diff
@@ -180,79 +97,166 @@ module.exports = function (opts) {
+        if (isUpsertRequested(msg)) {
+          return doUpsert(msg, coll, done)
+        }
+
+        return createNew(msg, coll, done)
 
-        var entp = ent.data$(false)
 
-        if (!update) {
-          var id
-          if (undefined !== ent.id$) {
-            id = ent.id$
-          } else if (opts.generate_id) {
-            id = opts.generate_id(ent)
+        function isUpsertRequested(msg) {
+          if (null == msg.q) {
+            return false
           }
 
-          entp._id = makeid(id)
+          if (!Array.isArray(msg.q.upsert$)) {
+            return false
+          }
 
-          coll.insertOne(entp, function (err, inserts) {
-            if (!error(args, err, cb)) {
-              var entu = inserts.ops[0]
-              var fent = null
-              if (entu) {
-                entu.id = idstr(entu._id)
-                delete entu._id
-                //fent = ent.make$(_.cloneDeep(entu))
-                fent = ent.make$(seneca.util.deep(entu))
+          const upsert_fields = clean_array(msg.q.upsert$)
+          const public_entdata = msg.ent.data$(false)
+
+          return upsert_fields.length > 0 &&
+            upsert_fields.every(p => p in public_entdata)
+        }
+
+        function doUpsert(msg, coll, done) {
+          const upsert_fields = clean_array(msg.q.upsert$)
+          const public_entdata = msg.ent.data$(false)
+
+          const filter_by = upsert_fields
+            .reduce((acc, field) => {
+              acc[field] = msg.ent[field]
+              return acc
+            }, {})
+
+          const replacement = (() => {
+            const o = Dot.flatten(Object.assign({}, public_entdata))
+            const id = ensure_id(msg.ent, opts)
+
+            if (null != id) {
+              o.$setOnInsert = { _id: id }
+            }
+
+            return o
+          })()
+
+          return coll.findOneAndUpdate(
+            filter_by,
+            replacement,
+            { upsert: true, returnNewDocument: true },
+
+            function (err, update) {
+              if (error(msg, err, done)) {
+                return
               }
-              seneca.log.debug('save/insert', ent, desc)
-              cb(null, fent)
+
+              const doc = update.value
+              const fent = makeEntityOfDocument(doc, msg.ent)
+
+              seneca.log.debug('save/upsert', msg.ent, desc)
+
+              return done(null, fent)
             }
-          })
-        } else {
-          var q = { _id: makeid(ent.id) }
-          delete entp.id
+          )
+        }
 
-          var shouldMerge = true
-          if (opts.merge !== false && ent.merge$ === false) {
-            shouldMerge = false
-          }
-          if (opts.merge === false && ent.merge$ !== true) {
-            shouldMerge = false
-          }
 
-          var set = entp
-          var func = 'replaceOne'
+        function createNew(msg, coll, done) {
+          const new_doc = (function () {
```

Comment:
why is this an IIFE?

Reply:
This is an IIFE for the following purposes:
- grouping code
- if deemed necessary, this approach lends itself to simpler refactoring
- makes it easy to find identical code throughout the code base and DRY it up
- better readability - when you read code that looks like this, it is easier to mentally separate the intention that each group of code tries to convey
- avoids mixing up logic

One might ask: why not just extract it into a separate function? I would argue that extraction of code into a separate function is not always a good idea. In situations when certain logic only appears once throughout the code base - such refactoring would in fact make it harder to understand the code.